text-generation-inference/server/text_generation_server/generator.py

```diff
@@ -337,45 +355,59 @@ def prefill(self, batch: Batch) -> Tuple[List[Generation], CachedBatch]:
             # Dynamically create a new slot for each request
             slot = Slot(self.slot_index, self.tokenizer, self.model.device)
             self.slot_index += 1
-            slot.assign(request, self.model.generation_config)
+            slot.assign(self.batch_id, request, self.model.generation_config)
             self.slots.append(slot)
             logger.debug(f"Request {slot.request_id} assigned to slot {slot.id}")
         # Reconstruct the full inputs (without padding) as seen by the model.
         # This comprises:
         # - the inputs for new requests,
         # - the inputs and the generated text that has already been cached (i.e. excluding the last generated token)
         #   for unfinished requests.
-        inputs = [slot.cached_text for slot in self.slots]
-        # Tokenize with padding
-        padded_inputs = self.tokenizer(inputs, return_tensors="pt", padding=True).to(self.model.device)
-        #  If needed truncate sequences to fit into the static dimensions
-        seq_length = min(padded_inputs.input_ids.shape[-1], self.model.config.sequence_length)
-        input_ids = padded_inputs.input_ids[:, :seq_length]
-        attention_mask = padded_inputs.attention_mask[:, :seq_length]
+
+        # Prepare inputs. They need to be tokenized and truncated afterwards.
+        tokenized_inputs = []
+        seq_length = 0
+        for _, slot in enumerate(self.slots):
+            cur_input = slot.cached_text
+            tokenized_input = self.tokenizer(cur_input, return_tensors="pt")
+            # Truncate the input to truncation in the slot (coming from request) and the maximum sequence length
+            tokenized_input.input_ids = tokenized_input.input_ids[:, -slot.truncate:][:self.model.config.sequence_length]
+            tokenized_input.attention_mask = tokenized_input.attention_mask[:, -slot.truncate:][:self.model.config.sequence_length]
+            tokenized_inputs.append(tokenized_input)
+            # Update the maximum sequence length
+            seq_length = max(seq_length, tokenized_input.input_ids.size(-1))
+
+        batch_size = len(self.slots)
+        # Initialize input_ids and attention_mask with padding (to make them all the same size)
+        input_ids = torch.full((batch_size, seq_length), self.tokenizer.pad_token_id, dtype=torch.int64)
+        attention_mask = torch.full((batch_size, seq_length), 0, dtype=torch.int64)
```

Comment:
qq: Can't we make it use `int32` for `input_ids` and `attention_mask` ?

Reply:
Umh, maybe we could, but the price would be a frequent cast between int32 and int64, because PB integers are serialized in int64. I think it's not worth doing it here, we could look at the advantages/disadvantages in another PR.